packages/open-next/src/adapters/routing/matcher.ts

```diff
@@ -0,0 +1,182 @@
+        return (
+          cookies?.[redirect.key] !== "" &&
+          new RegExp(redirect.value ?? "").test(cookies[redirect.key] ?? "")
+        );
+      case "query":
+        return query[redirect.key] && Array.isArray(redirect.value)
+          ? redirect.value.reduce(
+              (prev, current) =>
+                prev || new RegExp(current).test(query[redirect.key] as string),
+              false,
+            )
+          : new RegExp(redirect.value ?? "").test(
+              (query[redirect.key] as string | undefined) ?? "",
+            );
+      case "host":
+        return (
+          headers?.host !== "" &&
+          new RegExp(redirect.value ?? "").test(headers.host)
+        );
+      default:
+        return false;
+    }
+  };
+
+function checkHas(
+  matcher: ReturnType<typeof routeHasMatcher>,
+  has?: RouteHas[],
+  inverted = false,
+) {
+  return has
+    ? has.reduce((acc, cur) => {
+        if (acc === false) return false;
+        return inverted ? !matcher(cur) : matcher(cur);
+      }, true)
+    : true;
+}
+
+export function addNextConfigHeaders(
+  event: InternalEvent,
+  requestHeaders: Record<string, string | undefined>,
+  configHeaders?: Header[] | undefined,
+) {
+  if (!configHeaders) return;
+  const { rawPath, headers, query, cookies } = event;
+  const matcher = routeHasMatcher(headers, cookies, query);
+
+  for (const { source, headers, has, missing } of configHeaders) {
+    if (
+      new RegExp(source).test(rawPath) &&
+      checkHas(matcher, has) &&
+      checkHas(matcher, missing, true)
+    ) {
+      headers.forEach((h) => {
+        requestHeaders[h.key] = h.value;
+      });
+    }
+  }
+}
+
+export function handleRewrites<T extends RewriteDefinition>(
+  event: InternalEvent,
+  rewrites: T[],
+) {
+  const { rawPath, headers, query, cookies } = event;
+  const matcher = routeHasMatcher(headers, cookies, query);
+  const rewrite = rewrites.find(
+    (route) =>
+      new RegExp(route.regex).test(rawPath) &&
+      checkHas(matcher, route.has) &&
+      checkHas(matcher, route.missing, true),
+  );
+
+  const urlQueryString = new URLSearchParams(convertQuery(query)).toString();
+  let rewrittenUrl = rawPath;
+  const isExternalRewrite = isExternal(rewrite?.destination);
+  debug("isExternalRewrite", isExternalRewrite);
+  if (rewrite) {
+    const { pathname, protocol, hostname } = getUrlParts(
+      rewrite.destination,
+      isExternalRewrite,
+    );
+    const toDestination = compile(escapeRegex(pathname ?? "") ?? "");
+    const fromSource = match(escapeRegex(rewrite?.source) ?? "");
+    const _match = fromSource(rawPath);
+    if (_match) {
+      const { params } = _match;
+      const isUsingParams = Object.keys(params).length > 0;
+      if (isUsingParams) {
+        const rewrittenPath = unescapeRegex(toDestination(params));
+        rewrittenUrl = isExternalRewrite
+          ? `${protocol}//${hostname}${rewrittenPath}`
+          : `${rewrittenPath}`;
```

Comment:
@conico974 the leading `/` here was causing the infinite 308 for `@modal/(.)[dynamic][stuff]`
Not sure if this causing `rewrittenPath` to be naked w/o the leading slash though.
We may want to keep the prefix slash there and do:
`rewrittenPath = rewrittenPath.replace(/\/{2}/g, '/')